packages/sanity/src/core/form/types/definitionExtensions.ts

```diff
@@ -0,0 +1,254 @@
+import {ComponentType} from 'react'
+import {
+  CrossDatasetReference,
+  FileValue,
+  Geopoint,
+  ImageValue,
+  ReferenceValue,
+  SlugValue,
+} from '@sanity/types'
+import {PreviewProps} from '../../components'
+import {
+  ArrayFieldProps,
+  ArrayOfPrimitivesFieldProps,
+  BooleanFieldProps,
+  NumberFieldProps,
+  ObjectFieldProps,
+  StringFieldProps,
+} from './fieldProps'
+import {
+  ArrayOfObjectsInputProps,
+  ArrayOfPrimitivesInputProps,
+  BooleanInputProps,
+  NumberInputProps,
+  ObjectInputProps,
+  StringInputProps,
+} from './inputProps'
+import {ObjectItem, ObjectItemProps, PrimitiveItemProps} from './itemProps'
+
+export interface ArrayOfObjectsComponents {
```

Comment:
A lot of these are identical, but having per-type explicit definitions feels more correct.